beacon-chain/state/stategen/migrate.go

```diff
@@ -14,143 +12,54 @@ import (
 // MigrateToCold advances the split point in between the cold and hot state sections.
 // It moves the recent finalized states from the hot section to the cold section and
 // only preserve the ones that's on archived point.
-func (s *State) MigrateToCold(ctx context.Context, finalizedSlot uint64, finalizedRoot [32]byte) error {
+func (s *State) MigrateToCold(ctx context.Context, fSlot uint64, fRoot [32]byte) error {
 	ctx, span := trace.StartSpan(ctx, "stateGen.MigrateToCold")
 	defer span.End()
 
-	// Verify migration is sensible. The new finalized point must increase the current split slot, and
-	// on an epoch boundary for hot state summary scheme to work.
-	currentSplitSlot := s.splitInfo.slot
-	if currentSplitSlot > finalizedSlot {
+	lastFSlot := s.splitInfo.slot
+	lastFRoot := s.splitInfo.root
+	if lastFSlot > fSlot {
 		return nil
 	}
 
-	// Migrate all state summary objects from cache to DB.
-	if err := s.beaconDB.SaveStateSummaries(ctx, s.stateSummaryCache.GetAll()); err != nil {
-		return err
-	}
-	s.stateSummaryCache.Clear()
-
-	lastArchivedIndex, err := s.beaconDB.LastArchivedIndex(ctx)
-	if err != nil {
-		return err
-	}
-
-	// Move the states between split slot to finalized slot from hot section to the cold section.
-	filter := filters.NewFilter().SetStartSlot(currentSplitSlot).SetEndSlot(finalizedSlot - 1)
-	blockRoots, err := s.beaconDB.BlockRoots(ctx, filter)
-	if err != nil {
-		return err
-	}
-
-	for _, r := range blockRoots {
-		stateSummary, err := s.beaconDB.StateSummary(ctx, r)
-		if err != nil {
-			return err
-		}
-		if stateSummary == nil || stateSummary.Slot == 0 {
-			continue
-		}
-
-		archivedPointIndex := stateSummary.Slot / s.slotsPerArchivedPoint
-		nextArchivedPointSlot := (lastArchivedIndex + 1) * s.slotsPerArchivedPoint
-		// Only migrate if current slot is equal to or greater than next archived point slot.
-		if stateSummary.Slot >= nextArchivedPointSlot {
-			// If was a skipped archival index. The node should recover previous last archived index and state.
-			if skippedArchivedPoint(archivedPointIndex, lastArchivedIndex) {
-				recoveredIndex, err := s.recoverArchivedPoint(ctx, archivedPointIndex)
-				if err != nil {
-					return err
-				}
-				lastArchivedIndex = recoveredIndex
-			}
-
-			if !s.beaconDB.HasState(ctx, r) {
-				continue
-			}
-			if err := s.beaconDB.SaveArchivedPointRoot(ctx, r, archivedPointIndex); err != nil {
+	s.finalized.lock.RLock()
+	for i := lastFSlot; i < fSlot; i++ {
+		if i%s.slotsPerArchivedPoint == 0 && i != 0 {
+			aIndex := i / s.slotsPerArchivedPoint
+			if err := s.beaconDB.SaveState(ctx, s.finalized.state, lastFRoot); err != nil {
```

Comment:
Why do you save this state (potentially) multiple times?

Reply:
This is not right. We'll need to fix the edge case where multiple finalized check points got skipped. 
The FIFO cache will come in handy for this

Reply:
Do you need this loop at all? 

Maybe we can clarify the constraints:
- The gap between lastFSlot and fSlot will always be in the range of `s.slotsPerArchivedPoint` to `2*s.slotsPerArchivedPoint-1`

Is this true? Maybe need to add some validation or you could find that this code saves the same state 2 or more times.

Are you 100% sure that `lastFRoot` is the correct root? I thought it would have been `fRoot`. If it's not `fRoot` then what is `fRoot`?

Reply:
Actually maybe it needs to be in the range of 0 to s.slotsPerArchivedPoint?

Reply:
It would not be `fRoot`. `fRoot` is the new finalized root. 
`lastFRoot` (old) is the one that gets migrated from cold to hot. I verified and confirmed this is correct syncing Onyx.
I'll probably rename `lastFRoot` to `oldFRoot` or `prevCachedRoot` or something better...

I dont think this loop will work during skipped finalized check points. We have to use some sort of loop to handle this.

Will have to think about this more